### DISCUSSION TITLE
How to make sure the minimun scale larger than 0.001 in two-scale 3D Tess?

### DISCUSSION THREAD

**Discussion by mqc007** [neperfepx/neper#574] [Q&A] [answered]
Hi,
I'm trying to build a tess with two-scale in a cylinder (0.050,0.100). I hope it can contain more 4000 cells. 
so here is my code
 
`neper -T -n 1000::4 -domain "cylinder(0.1,0.05)" -morpho gg -reg 1 -fmax 0 -rsel 3 -group "(scaleid(1))" -statcell diameq -format tess,3dec -o c1000x4`

the point is i have to know the exactly two numbers of the two-scale for example 1000::4, because i want to group the cells in 3DEC.
when i use the V-Module to picture the tessellation i found that the minimum scale is about 0.0005. the maxmum is larger than 0.013.
another thing is, if the scale of the minimun cell is less than 0.001 it will be very hard to continue my analyse in 3DEC.

so is there any possible to make sure the minimum scale larger than 0.001 under these conditions?

best regards
many thanks

**Comment by rquey** [comment, marked as answer]
First thing first, there is no reason why you would not be able to group cells (for 3DEC) while having different numbers of subcells from one primary cell to another.

For example:
```
neper -T -n 100::1000*area -dim 2 -morpho gg -group "scaleid(1)" -o test
neper -V test.tess -datacellcol group -print img
```
![image](https://user-images.githubusercontent.com/2659376/194482987-4ffb5c00-fd6e-46f9-8a89-5b9744771572.png)
In this example, the number of subcells of a cell is proportional to its surface area.

You can even generate equal-size subcells (`centroidal` is a fast way to do it):
![image](https://user-images.githubusercontent.com/2659376/194483534-483cbc8f-6a4b-403d-9656-e528f510ae70.png)

Both examples will clearly lead to cell sizes that are closer to each other than what you get with `1000::4`.

(Also note that `-reg 1 -fmax 0` is *almost* equivalent to no regularization at all).

**Reply by mqc007** [reply]
thank you sir
what i am doing is about rock mechanics

**Reply by mqc007** [reply]
thank you sir
the format .3dec files is only for 3DEC 5.2 version 
when i use the 3DEC7.0 version i need to transform the code in the language of 3DEC7.0 and after that the group function is no longer supported. 
![image](https://user-images.githubusercontent.com/41913110/194497364-8be75e1f-204c-48a2-8524-316107280186.png)

so I have to group the cells with fish function. for this reason i need to know how many sub cells in the primary cell. as a result i used 1000::4 method, so i can wirte a loop program to group subcells.

i thought maybe i can set a limit for the minimum size of the cells to make sure each cell is larger than 0.001. is there any relevent code about that? 

and thank you for your note. 
I've asked you before. `-fmax 0 ` is helpful to make sure every face is a plane.

**Reply by rquey** [reply]
You can always run `-statcell scaleid(1)"` to know how many subcells per cell there are:
```
$ neper -T -n 100::1000*area -dim 2 -morpho gg -group "scaleid(1)" -o test -statcell "scaleid(1)"
$ uniq -c test.stcell | awk '{print "Cell "$2": "$1" subcells"}' 
Cell 1: 22 subcells
Cell 2: 13 subcells
Cell 3: 12 subcells
Cell 4: 26 subcells
Cell 5: 9 subcells
Cell 6: 5 subcells
Cell 7: 25 subcells
Cell 8: 36 subcells
Cell 9: 16 subcells
Cell 10: 22 subcells
...
```

In principle, you can use `diameq:interval(0.001,1)` so that no cell size is smaller than 0.001, but it won't necessarily be possible with `-n 1000::4` (suppose the primary cell itself is smaller than 0.001).

**Reply by mqc007** [reply]
many thanks. 
I will try

**Reply by mqc007** [reply]
@rquey Sir, I did try with the code  `diameq:interval(0.001,1)` but I went something wrong, the minimum sacle is still samller than 0.001
here is my code
`neper -T -n 1000::4 -morpho "diameq:interval (0.002,0.012)" -domain "cylinder(0.1,0.05)" -morpho gg -reg 1 -fmax 0 -rsel 3 -group "(scaleid(1))" -statcell diameq -format tess,3dec -o 1000x4new`
and V module code `neper -V 1000x4new.tess -datacellcol "real:file(1000x4new.stcell)" -datacellscale 0.001:0.003:0.005:0.007:0.009:0.011:0.013 -imagesize 1920:1080 -datacellscaletitle "Cell size(m)" -datacelltrs 0.5 -print 1000x4new`
![image](https://user-images.githubusercontent.com/41913110/194529718-655166b7-274f-4747-ac33-866f068b2714.png)

Sorry to bother you.

**Reply by rquey** [reply]
Please copy text as raw text, not images...

You have to specify `diameq:interval(0.002,0.012)` at scale 2, so, assuming you would keep `gg` at scale 1: `-morpho "gg:diameq:interval(0.002,0.012)"`.

**Reply by mqc007** [reply]
right. of course i need to specify two scales. thank you very much!

**Reply by mqc007** [reply]
@rquey Sir, I'm back. the code doesn't work, it says i have discovered a bug in neper.
`Info   :   [    2]  797/800 Info   :   - Setting seeds... 100%          
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =0.974178833
Info   :     > Iteration    491: fmin=0.793700527 f=0.793700535  
Info   :     > Final solution  : f   =0.793700527 (491 iterations)
Info   :     > Reached `eps' criterion.
Info   :   [    2]  798/800 Info   :   - Setting seeds... 100%          
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =0.961398256
Info   :     > Iteration    566: fmin=0.793700526 f=0.793700532  
Info   :     > Final solution  : f   =0.793700526 (566 iterations)
Info   :     > Reached `eps' criterion.
Info   :   [    2]  799/800 Info   :   - Setting seeds... 100%          
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =0.994133089
Info   :     > Iteration    525: fmin=0.793700574 f=0.793700645  
Info   :     > Final solution  : f   =0.793700574 (525 iterations)
Info   :     > Reached `eps' criterion.
Info   :   [    2]  800/800 Info   :   - Setting seeds... 100%          
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   :     > Initial solution: f   =0.993893260
Info   :     > Iteration    329: fmin=0.793701397 f=0.793701486  
Info   :     > Final solution  : f   =0.793701397 (329 iterations)
Info   :     > Reached `eps' criterion.
Info   : Flattening 2-scale tessellation...
Info   :   [level] Flattening 2-scale tessellation...
Info   :   [    1]  1/1: id=1                                           
Error  : You have discovered a bug in Neper!  Please file an issue at
         https://github.com/rquey/neper/issues.  Thank you.

`

I tied 1000::4 and 800::4, both failed.
the code i used:
`neper -T -n 800::4 -morpho "gg::diameq:interval (0.001,0.10)" -domain "cylinder(0.1,0.05)" -reg 1 -fmax 0 -rsel 3 -group "(scaleid(1))" -statcell diameq -format tess,3dec -o 1000x4new
`
